### PR TITLE
ENH: uniformize API for "center" argument for plot windows

### PR DESCRIPTION
**Pull request by neutrinoceros**
## PR Summary

Original goal: make errors much more informative than the current `RuntimeError(f'center keyword "{center}" not recognized')` raised by `PlotWindows.sanitize_center`, and improve the discoverability of the rich API that's actually implemented via error messages.
I've extended the goal to providing a more uniform API between plot windows and data selectors (which also have a "center" argument), because there was a lot of similar-yet different code basically accomplishing the same task there, as well as a couple things that worked with one but not the other.

I've added extensive tests for `CoordinateHandler.sanitize_center` and made the following changes:
- add much more verbose error message in case of invalid data type or incorrect value
- invalidate `center=YTArray(...)` if the array has an incorrect shape or unit dimension (previously any YTArray would pass)
- ~deprecate `center="c"` and `center="m"` because this encourages user code that's less clear, and reserving `"m"` for "max" doesn't leave any room for an equivalent "min" shortened keyword.~ edit: withdrawn
- ~deprecate `center="max"` (where the `("gas", "density)` field is implicitly queried) because it's not explicit as well as brittle (not every dataset has a `("gas", "density")` field)~ edit: withdrawn
- ~updated docs and tests to avoid showing off deprecated usages~ edit: withdrawn

I'm also using this to experiment with new possible special values, akin to `center="center"`, such as `center="left"` and `center="right"`, which I'm sure are useful to easily visualise domain boundaries / box "faces". This may be separated from this PR depending on how #3828 evolves.

## PR Checklist

- [x] Tests for invalid user input
- [x] Tests for deprecated user input
- [x] Tests for valid user input
- [x] Document `center="left"` and `center="right"`
- [x] Uniformize API for `yt.funcs.validate_center` and `YTDataContainer._set_center` (with minimal or no breaking changes)
- [x] make sure docstrings are in sync and up to date

### CODE REVIEW

**Comment by neutrinoceros** [comment]
There's one regression:
```
pytest yt/visualization/tests/test_particle_plot.py::TestParticleProjectionPlotSave::test_particle_plot_c -x
```
This test is failing because the dataset produced with `fake_particle_ds` doens't have a `"code_length"` registered. I don't know if this is an intended property of particle datasets but I doubt it. Right now I'm assuming this is really a bug revealed by my refactor.

**Comment by neutrinoceros** [comment]
there's one more regression in a test that I'm not sure is correct : `yt/data_objects/tests/test_covering_grid.py::test_covering_grid`
In this test, we create a sphere selection of a cylindrical dataset, with `center=(0, 0, 0)`, but the domain only goes from 1 to 2 (radially), so it's failing because the selected center is outside the domain. This behaviour is intentional on my part, though I'm not sure what the test's original author meant to check there. I think there is very little chance that a sphere selection with any center ever worked on cylindrical dataset (at least not if we mean a sphere in the cartesian space) anyway, so I'll probably end up changing the test, but I need to check what's the actual result as of the current tip of the main branch first.

**Comment by matthewturk** [comment]
I am sorry for not seeing this sooner, but I do not think we should deprecate `center="c"` or `center="max"`.  The other changes I think are fine, and I think an argument could be made to have a "default" field other than density specified on a per-frontend basis, but I am not in favor of this change -- which is a *pretty big change*!

**Comment by neutrinoceros** [comment]
Yeah I figured the deprecation part wouldn't go down easy... and this Pr has grown orders of magnitude beyond the original goal. If I drop the deprecation can we still move forward with the more consensual "api uniformity" part ?

**Comment by matthewturk** [comment]
That sounds like a good plan.  Also, if you knew it wouldn't go down, were you just testing boundaries?!?!?!?!  ;-)

**Comment by neutrinoceros** [comment]
> were you just testing boundaries?!?!?!?! ;-)

Well, that, but also trying to remove a few balls so the juggling may turn a bit easier instead for a change. Also it felt in line with the « no ambiguous fields » stuff we have rolling.
Any way I’ll update this later, which will also make it a lot easier to review, I’m hoping.

**Comment by neutrinoceros** [comment]
Btw how about deprecating the ˋ"max_density"ˋ string magic instead ? Would that be ok ? I think it's not needed since we support passing `("max", "density")` _and_ `("max", ("gas", "density"))` instead, which is more explicit/scalable/flexible/maintainable, and probably more discoverable by similarity with the rest of the public API ?

**Comment by matthewturk** [comment]
As a note about selections -- selections are always done in cartesian coordinates in *index* space, which is really confusing.  So for instance, a `region` in a spherical dataset is really a wedge, as it selects in [r, theta, phi] space, not [x, y, z] space.

**Comment by neutrinoceros** [comment]
I think the docs build failure is solved in #3965

**Comment by neutrinoceros** [comment]
@yt-fido test this please

**Comment by neutrinoceros** [comment]
This is now rebased on top of #4018 (previously I implemented an equivalent but local solution here)

**Comment by neutrinoceros** [comment]
As much as I would like this to go in for yt 4.1, it is by no means a blocker, so I'll remove it from the milestone

**Comment by jzuhone** [comment]
Sorry, wanted to look at this, but there's simply not enough time for all of the PRs that are out there.

**Comment by neutrinoceros** [comment]
No worries John, I'm just doing some housecleaning to resolve conflicts, and needed to close/reopen to refresh CI.

**Comment by neutrinoceros** [comment]
@matthewturk and @jzuhone, just letting you know that I have time to go through reviews these days, in case you guys are still interested in this one.

**Review by chrishavlin — COMMENTED** [review]
This generally looks good to me, I left some pretty minor comments. 

Do you mind updating the PR title? The present version of the PR really just touches the plotting containers: the data selectors are still using `yt.funcs.validate_center`.

File: doc/source/visualizing/plots.rst [hunk 1]
```diff
@@ -164,13 +164,17 @@ If supplied without units, the center is assumed by in code units.  There are al
 the following alternative options for the ``center`` keyword:
 
 * ``"center"``, ``"c"``: the domain center
+* ``"left"``, ``"l"``, ``"right"`` ``"r"``: the domain's left/right edge along the normal direction
```
[the remark this answers is not in the document]
Comment (chrishavlin) [reply doc/source/visualizing/plots.rst:167]: oh, whoops, forgot to come back to this. looking at the code, I understand now.

[the remark this answers is not in the document]
Comment (chrishavlin) [reply doc/source/visualizing/plots.rst:167]: though the fact that I had to look at the code to understand it maybe implies the wording here could be improved? I'm just not sure how...

[the remark this answers is not in the document]
Comment (neutrinoceros) [reply doc/source/visualizing/plots.rst:167]: Maybe I could specify that by "normal", I'm referring to `yt.SlicePlot`'s second argument (which I think is most often used _positionally_ in docs, so you are right that this won't be obvious to most readers)

File: yt/funcs.py [hunk 2]
```diff
@@ -1222,6 +1231,118 @@ def validate_center(center):
...
+def parse_center_array(center, ds, axis: Optional[int] = None):
+    known_shortnames = {"m": "max", "c": "center", "l": "left", "r": "right"}
+    valid_single_str_values = ("center", "left", "right")
+    valid_field_loc_str_values = ("min", "max")
+    valid_str_values = valid_single_str_values + valid_field_loc_str_values
+    default_error_message = (
+        "Expected any of the following\n"
+        "- 'c', 'center', 'l', 'left', 'r', 'right', 'm', 'max', or 'min'\n"
+        "- 'min' or 'max', followed by a field identifier\n"
+        "- a 3 element unyt_array with length dimensions"
```
[the remark this answers is not in the document]
Comment (neutrinoceros) [reply yt/funcs.py:1243]: will be done in my next commit

File: yt/visualization/fits_image.py [hunk 3]
```diff
@@ -974,19 +974,31 @@ class FITSSlice(FITSImageData):
...
-        The coordinate of the center of the image. If set to 'c', 'center' or
-        left blank, the plot is centered on the middle of the domain. If set
-        to 'max' or 'm', the center will be located at the maximum of the
-        ('gas', 'density') field. Centering on the max or min of a specific
-        field is supported by providing a tuple such as ("min","temperature")
-        or ("max","dark_matter_density"). Units can be specified by passing in
-        *center* as a tuple containing a coordinate and string unit name or by
-        passing in a YTArray. If a list or unitless array is supplied, code
-        units are assumed.
+    center : 'center', 'c', 'left', 'l', 'right', 'r' id of a global extremum, or array-like
```
[the remark this answers is not in the document]
Comment (neutrinoceros) [reply yt/visualization/fits_image.py:977]: done

File: yt/visualization/plot_modifications.py [hunk 4]
```diff
@@ -159,7 +159,7 @@ def _project_coords(self, plot, coord):
             # if this is an off-axis project or slice (ie cutting plane)
             # we have to calculate where the data coords fall in the projected
             # plane
-            elif ax == 4:
+            elif ax is None:
```
[the remark this answers is not in the document]
Comment (neutrinoceros) [reply yt/visualization/plot_modifications.py:162]: oh, great catch ! Indeed the `if/elif/else` structure is now redundant, I failed to see that, thanks !

File: yt/visualization/plot_window.py [hunk 5]
```diff
@@ -2527,16 +2576,26 @@ def plot_2d(
...
-         The coordinate of the center of the image. If set to 'c', 'center' or
-         left blank, the plot is centered on the middle of the domain. If set to
-         'max' or 'm', the center will be located at the maximum of the
-         ('gas', 'density') field. Centering on the max or min of a specific
-         field is supported by providing a tuple such as ("min","temperature") or
-         ("max","dark_matter_density"). Units can be specified by passing in *center*
-         as a tuple containing a coordinate and string unit name or by passing
-         in a YTArray. If a list or unitless array is supplied, code units are
-         assumed. For plot_2d, this keyword accepts a coordinate in two dimensions.
+    center : 'center', 'c', id of a global extremum, or array-like
```
[the remark this answers is not in the document]
Comment (neutrinoceros) [reply yt/visualization/plot_window.py:2579]: so projects do that (I'm pretty sure Matplotlib does), and indeed it'd would help here

File: yt/visualization/tests/test_plotwindow.py [hunk 6]
```diff
@@ -83,13 +83,17 @@ def setup():
     "M",
     "max",
     "Max",
-    "c",
-    "C",
+    "min",
+    "Min",
```
[the remark this answers is not in the document]
Comment (neutrinoceros) [reply yt/visualization/tests/test_plotwindow.py:87]: good call, reverted !

File: doc/source/visualizing/plots.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (chrishavlin) [reply doc/source/visualizing/plots.rst:167]: How about also specifying what happens for the other dimensions? e.g., "the domain's left/right edge along the normal direction. Remaining axes use their respective domain center values" . I think that would have been clearer to me initially.

[the remark this answers is not in the document]
Comment (chrishavlin) [reply doc/source/visualizing/plots.rst:167]: but I'm also OK with just leaving this. I don't honestly know why I was initially so confused by the wording.... I'll leave it up to you.

**Review by chrishavlin — DISMISSED** [review]

File: doc/source/visualizing/plots.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (neutrinoceros) [reply doc/source/visualizing/plots.rst:167]: I think it's worth being exhaustive, I'll implement your suggestion !

**Review by chrishavlin — APPROVED** [review]

**Comment by neutrinoceros** [comment]
@yt-fido test this please

**Comment by Xarthisius** [comment]
@yt-fido test this please

**Comment by neutrinoceros** [comment]
Thanks!
@matthewturk @jzuhone, do we need a second reviewer for this one ?

**Comment by matthewturk** [comment]
LGTM
